File: Research/backup/Analysis/heartbeat_monitor.py
#!/usr/bin/env python3
"""
Heartbeat Monitor - Send periodic heartbeats to GD32 and log responses

This script:
1. Sends heartbeat packets (0x06) every 50ms to keep GD32 alive
2. Captures and logs all responses from GD32
3. Parses sensor data packets (0x15) to show real-time sensor values
4. Runs for specified duration or until Ctrl+C

Usage: python3 heartbeat_monitor.py [duration_seconds]
"""

import serial
import time
import sys
import signal
from datetime import datetime

# Protocol constants
SYNC1 = 0xFA
SYNC2 = 0xFB
CMD_HEARTBEAT = 0x06
CMD_SENSOR = 0x15

# Global flag for clean shutdown
running = True

def signal_handler(sig, frame):
    """Handle Ctrl+C gracefully"""
    global running
    print("\n\n[*] Shutting down...")
    running = False

def build_heartbeat_packet():
    """Build heartbeat packet: FA FB 03 06 00 06"""
    # LENGTH = 3 (CMD + DATA + CRC)
    # CMD = 0x06
    # DATA = 0x00 (no data)
    # CRC = 0x06 (observed from captures)
    return bytes([SYNC1, SYNC2, 0x03, CMD_HEARTBEAT, 0x00, 0x06])

def parse_packet(data):
    """Parse received packet"""
    if len(data) < 4:
        return None

    if data[0] != SYNC1 or data[1] != SYNC2:
        return None

    length = data[2]
    cmd_id = data[3]

    expected_total = 3 + length  # SYNC1 + SYNC2 + LEN + (CMD + DATA + CRC)

    if len(data) < expected_total:
        return None

    if length > 0:
        crc = data[2 + length]
        payload = data[4:2 + length]
    else:
        crc = None
        payload = b''

    return {
        'cmd_id': cmd_id,
        'length': length,
        'payload': payload,
        'crc': crc,
        'raw': data[:expected_total]
    }

def parse_sensor_data(payload):
    """Parse sensor data payload (99 bytes based on captures)"""
    if len(payload) < 97:  # 99 - 1 CMD - 1 CRC = 97 data bytes
        return None

    # This is a guess based on typical robot sensor layouts
    # Will need to be refined by observing actual data changes
    try:
        import struct

        # Assuming little-endian int32 values at various offsets
        # These offsets are guesses and will be validated by observation

        sensor_info = {
            'timestamp_offset_0': struct.unpack('<I', payload[0:4])[0],
            'value_offset_4': struct.unpack('<I', payload[4:8])[0],
            'encoder_left?': struct.unpack('<I', payload[12:16])[0],
            'encoder_right?': struct.unpack('<I', payload[20:24])[0],
            'imu_accel_x?': struct.unpack('<h', payload[40:42])[0],
            'imu_accel_y?': struct.unpack('<h', payload[42:44])[0],
            'imu_accel_z?': struct.unpack('<h', payload[44:46])[0],
            'imu_gyro_x?': struct.unpack('<h', payload[46:48])[0],
            'imu_gyro_y?': struct.unpack('<h', payload[48:50])[0],
            'imu_gyro_z?': struct.unpack('<h', payload[50:52])[0],
            'battery?': payload[70:72].hex(),
            'bumper?': payload[80],
            'cliff?': payload[82],
        }
        return sensor_info
    except Exception as e:
        return None

def format_timestamp():
    """Get formatted timestamp"""
    return datetime.now().strftime("%H:%M:%S.%f")[:-3]

def main():
    global running

    # Parse arguments
    duration = int(sys.argv[1]) if len(sys.argv) > 1 else 60  # Default 60 seconds
    serial_port = '/dev/ttyS1'
    baud_rate = 115200

    print("=" * 80)
    print("GD32 Heartbeat Monitor")
    print("=" * 80)
    print(f"Serial Port: {serial_port}")
    print(f"Baud Rate: {baud_rate}")
    print(f"Duration: {duration} seconds")
    print(f"Heartbeat: Every 50ms")
    print()
    print("Instructions:")
    print("  - Script will keep GD32 alive with heartbeat packets")
    print("  - Move the robot manually to trigger sensors")
    print("  - Press bumpers, trigger cliff sensors")
    print("  - Spin wheels to see encoder changes")
    print("  - Press Ctrl+C to stop early")
    print()
    print("Starting in 3 seconds...")
    print("=" * 80)
    time.sleep(3)

    # Setup signal handler
    signal.signal(signal.SIGINT, signal_handler)

    # Open serial port
    try:
        ser = serial.Serial(
            port=serial_port,
            baudrate=baud_rate,
            bytesize=serial.EIGHTBITS,
            parity=serial.PARITY_NONE,
            stopbits=serial.STOPBITS_ONE,
            timeout=0.01  # 10ms timeout for non-blocking read
        )
    except Exception as e:
        print(f"[!] Failed to open {serial_port}: {e}")
        print("[!] Make sure AuxCtrl is stopped: killall AuxCtrl")
        return 1

    print(f"[+] Serial port opened: {serial_port}")

    # Prepare logging
    log_filename = f"/tmp/heartbeat_log_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
    logfile = open(log_filename, 'w')
    logfile.write(f"# GD32 Heartbeat Monitor Log\n")
    logfile.write(f"# Started: {datetime.now()}\n")
    logfile.write(f"# Format: [timestamp] direction command_id data_hex\n\n")

    print(f"[+] Logging to: {log_filename}")
    print()
    print("-" * 80)
    print("LIVE MONITORING (Ctrl+C to stop)")
    print("-" * 80)
    print()

    # Statistics
    tx_count = 0
    rx_count = 0
    sensor_packets = 0
    other_packets = 0
    errors = 0

    start_time = time.time()
    last_heartbeat = 0
    rx_buffer = bytearray()

    try:
        while running and (time.time() - start_time) < duration:
            current_time = time.time()

            # Send heartbeat every 50ms
            if (current_time - last_heartbeat) >= 0.05:
                packet = build_heartbeat_packet()
                ser.write(packet)
                tx_count += 1
                last_heartbeat = current_time

                # Log TX
                logfile.write(f"[{format_timestamp()}] TX {packet.hex(' ')}\n")

            # Read any incoming data
            if ser.in_waiting > 0:
                chunk = ser.read(ser.in_waiting)
                rx_buffer.extend(chunk)

                # Try to parse complete packets from buffer
                while len(rx_buffer) >= 4:
                    # Look for sync bytes
                    if rx_buffer[0] == SYNC1 and rx_buffer[1] == SYNC2:
                        packet = parse_packet(rx_buffer)
                        if packet:
                            rx_count += 1

                            # Log RX
                            logfile.write(f"[{format_timestamp()}] RX {packet['raw'].hex(' ')}\n")

                            # Process packet
                            if packet['cmd_id'] == CMD_SENSOR:
                                sensor_packets += 1
                                print(f"[{format_timestamp()}] SENSOR DATA (packet #{sensor_packets}):")

                                # Try to parse sensor fields
                                sensor_info = parse_sensor_data(packet['payload'])
                                if sensor_info:
                                    for key, value in sensor_info.items():
                                        print(f"  {key:20s}: {value}")
                                else:
                                    print(f"  Raw ({len(packet['payload'])} bytes): {packet['payload'].hex(' ')[:80]}...")
                                print()
                            else:
                                other_packets += 1
                                print(f"[{format_timestamp()}] CMD 0x{packet['cmd_id']:02X} "
                                      f"({len(packet['payload'])} bytes): {packet['raw'].hex(' ')}")

                            # Remove parsed packet from buffer
                            rx_buffer = rx_buffer[len(packet['raw']):]
                        else:
                            # Incomplete packet, wait for more data
                            break
                    else:
                        # Invalid sync, skip one byte
                        errors += 1
                        rx_buffer.pop(0)

            # Small sleep to prevent CPU spinning
            time.sleep(0.001)

    except Exception as e:
        print(f"\n[!] Error: {e}")
        import traceback
        traceback.print_exc()

    finally:
        # Cleanup
        ser.close()
        logfile.close()

        elapsed = time.time() - start_time

        print()
        print("=" * 80)
        print("SESSION SUMMARY")
        print("=" * 80)
        print(f"Duration: {elapsed:.1f} seconds")
        print(f"TX Heartbeats: {tx_count}")
        print(f"RX Packets: {rx_count}")
        print(f"  - Sensor data (0x15): {sensor_packets}")
        print(f"  - Other responses: {other_packets}")
        print(f"Sync errors: {errors}")
        print()
        print(f"Log saved to: {log_filename}")
        print()
        print("Next steps:")
        print(f"  - Analyze log: cat {log_filename}")
        print(f"  - Copy to computer: scp root@vacuum:{log_filename} .")
        print("=" * 80)

    return 0

if __name__ == '__main__':
    sys.exit(main())

File: Research/backup/Analysis/heartbeat_monitor.c
/*
 * GD32 Heartbeat Monitor
 *
 * Sends periodic heartbeat packets to GD32 and logs responses
 * Compile: gcc -o heartbeat_monitor heartbeat_monitor.c
 * Or for ARM: arm-linux-gnueabi-gcc -static -o heartbeat_monitor heartbeat_monitor.c
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include <time.h>
#include <signal.h>
#include <sys/time.h>

#define SERIAL_PORT "/dev/ttyS1"
#define BAUD_RATE B115200
#define HEARTBEAT_INTERVAL_MS 50

// Protocol constants
#define SYNC1 0xFA
#define SYNC2 0xFB
#define CMD_HEARTBEAT 0x06
#define CMD_SENSOR 0x15

// Global flag for clean shutdown
volatile int running = 1;

void signal_handler(int sig) {
    printf("\n[*] Shutting down...\n");
    running = 0;
}

long long current_time_ms() {
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return (long long)(tv.tv_sec) * 1000 + (tv.tv_usec) / 1000;
}

void print_timestamp() {
    struct timeval tv;
    gettimeofday(&tv, NULL);
    struct tm *tm_info = localtime(&tv.tv_sec);
    printf("[%02d:%02d:%02d.%03ld] ",
           tm_info->tm_hour, tm_info->tm_min, tm_info->tm_sec,
           tv.tv_usec / 1000);
}

void print_hex(const unsigned char *data, int len) {
    for (int i = 0; i < len; i++) {
        printf("%02x ", data[i]);
    }
}

int open_serial(const char *port) {
    int fd = open(port, O_RDWR | O_NOCTTY | O_NONBLOCK);
    if (fd < 0) {
        perror("open");
        return -1;
    }

    struct termios options;
    tcgetattr(fd, &options);

    // Set baud rate
    cfsetispeed(&options, BAUD_RATE);
    cfsetospeed(&options, BAUD_RATE);

    // 8N1, no flow control
    options.c_cflag &= ~PARENB;  // No parity
    options.c_cflag &= ~CSTOPB;  // 1 stop bit
    options.c_cflag &= ~CSIZE;
    options.c_cflag |= CS8;      // 8 data bits
    options.c_cflag |= (CLOCAL | CREAD);  // Enable receiver
    options.c_cflag &= ~CRTSCTS; // No hardware flow control

    // Raw input
    options.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
    options.c_iflag &= ~(IXON | IXOFF | IXANY);  // No software flow control
    options.c_oflag &= ~OPOST;  // Raw output

    // Set timeouts
    options.c_cc[VMIN] = 0;   // Non-blocking
    options.c_cc[VTIME] = 0;

    tcsetattr(fd, TCSANOW, &options);
    tcflush(fd, TCIOFLUSH);

    return fd;
}

void send_heartbeat(int fd, FILE *logfile) {
    // Heartbeat packet: FA FB 03 06 00 06
    unsigned char packet[] = {SYNC1, SYNC2, 0x03, CMD_HEARTBEAT, 0x00, 0x06};

    write(fd, packet, sizeof(packet));

    // Log TX
    if (logfile) {
        fprintf(logfile, "[");
        struct timeval tv;
        gettimeofday(&tv, NULL);
        fprintf(logfile, "%ld.%03ld", tv.tv_sec, tv.tv_usec / 1000);
        fprintf(logfile, "] TX ");
        for (int i = 0; i < sizeof(packet); i++) {
            fprintf(logfile, "%02x ", packet[i]);
        }
        fprintf(logfile, "\n");
        fflush(logfile);
    }
}

int parse_packet(const unsigned char *buffer, int buf_len, unsigned char *packet_out, int *packet_len) {
    if (buf_len < 4) return 0;

    if (buffer[0] != SYNC1 || buffer[1] != SYNC2) {
        return -1;  // Invalid sync
    }

    int length = buffer[2];
    int expected_total = 3 + length;  // SYNC1 + SYNC2 + LEN + payload

    if (buf_len < expected_total) {
        return 0;  // Incomplete packet
    }

    // Copy packet
    memcpy(packet_out, buffer, expected_total);
    *packet_len = expected_total;

    return expected_total;
}

void print_sensor_data(const unsigned char *payload, int len) {
    if (len < 97) {
        printf("  [Warning: Short payload %d bytes]\n", len);
        return;
    }

    // Print some key offsets (these are guesses based on captures)
    printf("  Offset 0-3    : %02x %02x %02x %02x\n",
           payload[0], payload[1], payload[2], payload[3]);
    printf("  Offset 12-15  : %02x %02x %02x %02x (encoder left?)\n",
           payload[12], payload[13], payload[14], payload[15]);
    printf("  Offset 20-23  : %02x %02x %02x %02x (encoder right?)\n",
           payload[20], payload[21], payload[22], payload[23]);
    printf("  Offset 40-45  : %02x %02x %02x %02x %02x %02x (IMU accel?)\n",
           payload[40], payload[41], payload[42], payload[43], payload[44], payload[45]);
    printf("  Offset 46-51  : %02x %02x %02x %02x %02x %02x (IMU gyro?)\n",
           payload[46], payload[47], payload[48], payload[49], payload[50], payload[51]);
    printf("  Offset 70-71  : %02x %02x (battery?)\n",
           payload[70], payload[71]);
    printf("  Offset 80     : %02x (bumper?)\n", payload[80]);
    printf("  Offset 82     : %02x (cliff?)\n", payload[82]);
}

int main(int argc, char *argv[]) {
    int duration = 60;  // Default 60 seconds

    if (argc > 1) {
        duration = atoi(argv[1]);
    }

    printf("================================================================================\n");
    printf("GD32 Heartbeat Monitor (C version)\n");
    printf("================================================================================\n");
    printf("Serial Port: %s\n", SERIAL_PORT);
    printf("Baud Rate: 115200\n");
    printf("Duration: %d seconds\n", duration);
    printf("Heartbeat: Every %d ms\n", HEARTBEAT_INTERVAL_MS);
    printf("\n");
    printf("Instructions:\n");
    printf("  - Move robot manually to trigger sensors\n");
    printf("  - Press bumpers, trigger cliff sensors\n");
    printf("  - Spin wheels to see encoder changes\n");
    printf("  - Press Ctrl+C to stop early\n");
    printf("\n");
    printf("Starting in 3 seconds...\n");
    printf("================================================================================\n");
    sleep(3);

    // Setup signal handler
    signal(SIGINT, signal_handler);

    // Open serial port
    int fd = open_serial(SERIAL_PORT);
    if (fd < 0) {
        printf("[!] Failed to open %s\n", SERIAL_PORT);
        printf("[!] Make sure AuxCtrl is stopped: killall AuxCtrl\n");
        return 1;
    }

    printf("[+] Serial port opened: %s\n", SERIAL_PORT);

    // Open log file
    char log_filename[256];
    time_t now = time(NULL);
    struct tm *tm_info = localtime(&now);
    snprintf(log_filename, sizeof(log_filename),
             "/tmp/heartbeat_log_%04d%02d%02d_%02d%02d%02d.log",
             tm_info->tm_year + 1900, tm_info->tm_mon + 1, tm_info->tm_mday,
             tm_info->tm_hour, tm_info->tm_min, tm_info->tm_sec);

    FILE *logfile = fopen(log_filename, "w");
    if (logfile) {
        fprintf(logfile, "# GD32 Heartbeat Monitor Log\n");
        fprintf(logfile, "# Format: [timestamp] direction hex_bytes\n\n");
        printf("[+] Logging to: %s\n", log_filename);
    }

    printf("\n");
    printf("--------------------------------------------------------------------------------\n");
    printf("LIVE MONITORING (Ctrl+C to stop)\n");
    printf("--------------------------------------------------------------------------------\n");
    printf("\n");

    // Statistics
    int tx_count = 0;
    int rx_count = 0;
    int sensor_packets = 0;
    int other_packets = 0;

    long long start_time = current_time_ms();
    long long last_heartbeat = 0;

    unsigned char rx_buffer[4096];
    int rx_buf_len = 0;

    while (running && (current_time_ms() - start_time) < duration * 1000) {
        long long current = current_time_ms();

        // Send heartbeat every 50ms
        if (current - last_heartbeat >= HEARTBEAT_INTERVAL_MS) {
            send_heartbeat(fd, logfile);
            tx_count++;
            last_heartbeat = current;
        }

        // Read incoming data
        unsigned char temp[256];
        int n = read(fd, temp, sizeof(temp));
        if (n > 0) {
            // Append to buffer
            if (rx_buf_len + n < sizeof(rx_buffer)) {
                memcpy(rx_buffer + rx_buf_len, temp, n);
                rx_buf_len += n;
            }

            // Try to parse packets from buffer
            while (rx_buf_len >= 4) {
                unsigned char packet[512];
                int packet_len = 0;

                int result = parse_packet(rx_buffer, rx_buf_len, packet, &packet_len);

                if (result > 0) {
                    // Valid packet
                    rx_count++;

                    // Log RX
                    if (logfile) {
                        fprintf(logfile, "[");
                        struct timeval tv;
                        gettimeofday(&tv, NULL);
                        fprintf(logfile, "%ld.%03ld", tv.tv_sec, tv.tv_usec / 1000);
                        fprintf(logfile, "] RX ");
                        for (int i = 0; i < packet_len; i++) {
                            fprintf(logfile, "%02x ", packet[i]);
                        }
                        fprintf(logfile, "\n");
                        fflush(logfile);
                    }

                    // Process packet
                    int cmd_id = packet[3];

                    if (cmd_id == CMD_SENSOR) {
                        sensor_packets++;
                        print_timestamp();
                        printf("SENSOR DATA (packet #%d):\n", sensor_packets);

                        int payload_len = packet[2] - 2;  // LEN - CMD - CRC
                        print_sensor_data(packet + 4, payload_len);
                        printf("\n");
                    } else {
                        other_packets++;
                        print_timestamp();
                        printf("CMD 0x%02x (%d bytes): ", cmd_id, packet_len);
                        print_hex(packet, packet_len);
                        printf("\n");
                    }

                    // Remove parsed packet from buffer
                    memmove(rx_buffer, rx_buffer + packet_len, rx_buf_len - packet_len);
                    rx_buf_len -= packet_len;

                } else if (result < 0) {
                    // Invalid sync, skip one byte
                    memmove(rx_buffer, rx_buffer + 1, rx_buf_len - 1);
                    rx_buf_len--;
                } else {
                    // Incomplete packet, need more data
                    break;
                }
            }
        }

        usleep(1000);  // 1ms sleep to prevent CPU spinning
    }

    // Cleanup
    close(fd);
    if (logfile) fclose(logfile);

    long long elapsed = current_time_ms() - start_time;

    printf("\n");
    printf("================================================================================\n");
    printf("SESSION SUMMARY\n");
    printf("================================================================================\n");
    printf("Duration: %.1f seconds\n", elapsed / 1000.0);
    printf("TX Heartbeats: %d\n", tx_count);
    printf("RX Packets: %d\n", rx_count);
    printf("  - Sensor data (0x15): %d\n", sensor_packets);
    printf("  - Other responses: %d\n", other_packets);
    printf("\n");
    printf("Log saved to: %s\n", log_filename);
    printf("\n");
    printf("Next steps:\n");
    printf("  - Analyze log: cat %s\n", log_filename);
    printf("  - Copy to computer: scp root@vacuum:%s .\n", log_filename);
    printf("================================================================================\n");

    return 0;
}
